OpInfo for `torch.nn.functional.normalize`

krshrimali opened the pull request:
See https://github.com/facebookresearch/functorch/issues/78 and https://github.com/pytorch/pytorch/issues/54261

cc: @mruberry @zou3519 @Chillee

krshrimali committed:
OpInfo for nn.functional.normalize

krshrimali committed:
Skip test_variant_consistency_jit

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2181,6 +2181,27 @@ def sample_inputs_max_min_binary(op_info, device, dtype, requires_grad, **kwargs
+        for input_shape, p, dim in cases:
+            yield SampleInput(make_arg(input_shape), args=(p, dim))
+        # No args case
+        yield SampleInput(make_arg((2, 3, 4, 5)))
```
These are good extra cases; a simpler way of writing this in the future might just be to add a dict to the tuples in `cases` above that was the kwargs. Most of them would be empty but that seems OK

krshrimali replied:
Thanks, @mruberry for taking a look. I've made the changes in the recent commit, I hope I didn't get your suggestion wrong? I'm sorry if I did.

One extra point, I've realized - since these aren't kw-only args, hence I would like to check their positions as well (`p` is floating-point and `dim` is integral type). So I've added a separate extra case while keeping `kwargs` in the `cases` tuple as a `dict`.

Do you think that's fine?

mruberry replied:
I appreciate your thinking and you could structure the cases like this:

```
cases = (
  (input, (args), {kwargs})
)
```

To cover both positional and keyword argument cases. Testing that a function accepts positional and keyword args correctly may not be that interesting, though, since it's testing the arg parser more than the function's logic. It might catch accidental regressions where an operation was made kwarg-only by mistake, but that seems like a rare case.

So I think the additional case is fine, but it can be modeled alongside the other cases, and it's probably not necessary. Does that make sense?

krshrimali replied:
Thanks, @mruberry! I agree, the positional arg testing isn't that interesting here. I removed this from the cases here, testing through `kwargs` should be enough here, I suppose.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6443,6 +6464,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+               # RuntimeError: aliasOp != torch::jit::getOperatorAliasMap().end()
+               # INTERNAL ASSERT FAILED at "../torch/csrc/jit/passes/utils/check_alias_annotation.cpp":159,
+               # please report a bug to PyTorch.
+               SkipInfo('TestJit', 'test_variant_consistency_jit',),
```
cc @eellison

krshrimali replied:
From what I understood so far, it's probably because `normalize` calls `norm` and `torch.div` (or `/` operator) in its definition: https://github.com/pytorch/pytorch/blob/88af4d84415855e0863a7138f87f96c203400efe/torch/nn/functional.py#L4437-L4441

Hence the graph created from: https://github.com/pytorch/pytorch/blob/88af4d84415855e0863a7138f87f96c203400efe/torch/testing/_internal/jit_metaprogramming_utils.py#L488-L491 doesn't consist of `aten::normalize` but instead has the nodes for `norm` and `div`. This leads to continuing from the loop here (no `if` condition satisfied): https://github.com/pytorch/pytorch/blob/88af4d84415855e0863a7138f87f96c203400efe/torch/csrc/jit/passes/utils/check_alias_annotation.cpp#L151-L153

And hence the final failure here: https://github.com/pytorch/pytorch/blob/88af4d84415855e0863a7138f87f96c203400efe/torch/csrc/jit/passes/utils/check_alias_annotation.cpp#L159

Was just trying to see what's happening. :) In case it helps!

mruberry commented:
@zou3519 would you make sure this covers the cases you need and shepherd this in?

krshrimali committed:
Bring the skip back...was removed by mistake

krshrimali committed:
Address review, minor addition to extra cases

krshrimali committed:
Fix, aten::normalize isn't correct, don't test for positional args

krshrimali committed:
Fix mypy failures

krshrimali committed:
Merge branch 'opinfo/high_priority/nn/functional/normalize' of https://github.com/krshrimali/pytorch into opinfo/high_priority/nn/functional/normalize

zou3519 commented:
Yes, can do!

krshrimali committed:
Merge branch 'master' into opinfo/high_priority/nn/functional/normalize

zou3519 commented:
@krshrimali could you rebase this please?

krshrimali committed:
Fix merge conflict, rebase

krshrimali committed:
supports_out is by default True, not needed here

krshrimali commented:
Thanks for the ping, @zou3519 - I've rebased the branch. Also removed `supports_out=True` since it's already set to `True` in the OpInfo class. Hope that's okay. Thanks for taking a look.

mruberry commented:
ping @zou3519

krshrimali committed:
Fix merge conflict

krshrimali committed:
Remove adaptive_avg_pool2d since it was reverted

krshrimali commented:
> Sorry, I missed this notification. @krshrimali could you please rebase this again? (it looks like there are merge conflicts)

Thanks, @zou3519 for taking a look. I understand, and I should have pinged again but I missed it as well. I have fixed the merge conflicts. :) Thanks again!
